packages/miniflare/src/shared/dev-registry.worker.ts

```diff
@@ -0,0 +1,349 @@
+	httpStyle?: "host" | "proxy";
+	path?: string;
+}
+
+const log = new Log();
+
+/**
+ * Worker thread implementation for dev registry with HTTP proxy server.
+ * This runs both the file watching and HTTP proxy in a separate thread
+ * to prevent blocking the main thread.
+ */
+class ProxyServer {
+	private registry: Record<string, WorkerDefinition> = {};
+	private runtimeEntryURL: string | null = null;
+	private fallbackServicePorts: Record<string, Record<string, number>> = {};
+	private subscribers: Map<string, Array<() => void>> = new Map();
+
+	constructor(private messagePort: MessagePort) {
+		this.setupMessageHandlers();
+	}
+
+	private setupMessageHandlers() {
+		this.messagePort.on("message", (message) => {
+			switch (message.type) {
+				case "update":
+					this.updateRegistry(message.workers);
+					break;
+				case "setup":
+					this.runtimeEntryURL = message.runtimeEntryURL;
+					this.fallbackServicePorts = message.fallbackServicePorts;
+					break;
+				default:
+					// Ignore unknown message types
+					break;
+			}
+		});
+	}
+
+	public async start() {
+		try {
+			// Listen on a random port
+			await new Promise<void>((resolve, reject) => {
+				const server = http.createServer({
+					// There might be no HOST header when proxying a fetch request made over service binding
+					//  e.g. env.MY_WORKER.fetch("https://example.com")
+					requireHostHeader: false,
+				});
+
+				server.on("request", this.handleRequest.bind(this));
+				server.on("connect", this.handleConnect.bind(this));
+				server.listen(0, "127.0.0.1", () => {
+					const address = server.address();
+
+					if (!address || typeof address === "string") {
+						reject(new Error("Failed to get server address"));
+						return;
+					}
+
+					// Notify parent that server is ready with the port
+					this.messagePort.postMessage({
+						type: "ready",
+						address: `${address.address}:${address.port}`,
+					});
+
+					resolve();
+				});
+				server.on("error", reject);
+			});
+		} catch (error) {
+			this.messagePort.postMessage({
+				type: "error",
+				error: error instanceof Error ? error.message : String(error),
+			});
+		}
+	}
+
+	private handleRequest(
+		req: http.IncomingMessage,
+		res: http.ServerResponse
+	): void {
+		const serviceProxyTarget = extractServiceFetchProxyTarget(req);
+		if (serviceProxyTarget) {
+			const address = this.getServiceAddress(
+				serviceProxyTarget.service,
+				serviceProxyTarget.entrypoint
+			);
+
+			this.handleProxy(req, res, address);
+			return;
+		}
+
+		const doProxyTarget = extractDoFetchProxyTarget(req);
+		if (doProxyTarget) {
+			const address = this.getDurableObjectAddress(
+				doProxyTarget.scriptName,
+				doProxyTarget.className
+			);
+
+			if (!address) {
+				res.writeHead(503);
+				res.end("Service Unavailable");
+				return;
+			}
+
+			this.handleProxy(req, res, address);
+			return;
+		}
+
+		// If no valid target found, return 404
+		res.writeHead(404);
+		res.end("Not Found");
+	}
```

Comment:
This was done in [`handleLoopback`](https://github.com/cloudflare/workers-sdk/pull/10142/files#r2252262548) before moving to the proxy server.